### PR TITLE
ARROW-6068: [C++] Allow passing Field instances to StructArray::Make

### PR DESCRIPTION
**Pull request by pitrou**
This helps fix a Python test failure when hypothesis testing is enabled.

### CODE REVIEW

File: cpp/src/arrow/array.h [hunk 1]
```diff
@@ -918,6 +918,17 @@ class ARROW_EXPORT StructArray : public Array {
       std::shared_ptr<Buffer> null_bitmap = NULLPTR,
       int64_t null_count = kUnknownNullCount, int64_t offset = 0);
 
+  /// \brief Return a StructArray from child arrays and fields.
+  ///
+  /// The length is automatically inferred from the arguments.
+  /// There should be at least one child array.  This method does not
+  /// check that field types and child array types are consistent.
```
Comment (bkietz) [inline cpp/src/arrow/array.h]: .. maybe it should?

**Review by bkietz — APPROVED** [review]

File: cpp/src/arrow/array.h (hunk 1 above)
Comment (wesm) [reply cpp/src/arrow/array.h]: I think that's the job of `Validate`

**Review by wesm — COMMENTED** [review]
Looks generally OK to me

File: cpp/src/arrow/array.h (hunk 1 above)
Comment (pitrou) [reply cpp/src/arrow/array.h]: Yes, because it's potentially expensive.

**Pull request closed without merging**